refactor: readMetadata & env.localization

SleepWalker opened the pull request:
<!--
Thank you for sending the PR! We appreciate you spending the time to work on these changes.

Help us understand your motivation by explaining why you decided to make this change.

You can learn more about contributing to Docusaurus here: https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md

Happy contributing!

-->

## Motivation

This PR appeared during during my work on #868. It contains no new features or logic changes. I've rearranged the code to decouple it from the environment and be able to cover it with tests.

The docs-related logic from `server/readMetadata` was moved into a separate class `DocsMetadata` and covered with test.

I've also removed translations.js and replaced it with `Translation.t(language, key, category)` interface and moved `getLanguage` implementation from utils to `Translation`. The `Translation` class was moved from `server/env` into a separate file

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

The best way is to compare `metadata.js` before and after my refactoring applied (e.g. with jsondiff.com). They should be the same. The problem is that I can not test with translations applied. I've compared only the `metadata.js` files I can get locally.

Another one option is to compare current gh-pages version and netlify version using different locales and doc versions.

SleepWalker commented:
@endiliey ok. I can restore that file, but probably it would be better to add a deprecation warning in that file?

If the user will access the data through the interface it will allow Docusaurus developers to safely change the way that data handled.

---

Regarding `Translation.t(language, key, category)`. I think that in future it will be possible to set the current locale for Translation, so the first argument may be optional in future. Considering this I can implement the logic that will support following signatures:

* `Translation.t(key, language)`
* `Translation.t(key, category, language)`

This will make internals of `Translate.t` more complicated, but in future we will be able to migrate to the following signatures:

* `Translation.t(key)`
* `Translation.t(key, category)`

JoelMarcey commented:
Yeah, we would definitely have to start with a deprecation and give a few releases and warnings in the release notes about it since if any project that uses translations would require the current file. But I think that is ok.

SleepWalker commented:
@endiliey the `translation.js` is back and exports `Translation` instance. I have also removed language argument to `translation.t`. Now it will use the language, that will be set before rendering.

When some one will try to use `translation[langId]` the warning will be printed in console.

I have merged with current master and added support for keys as array, that was introduced in #917, e.g.:

```js
    const title = translation.t(['docs', id, 'title']) || defaultTitle;
```

endiliey commented on lib/server/env/Translation.js:
```diff
@@ -0,0 +1,190 @@
+      .replace(new RegExp('\\\\', 'g'), '\\');
+  }
+
+  resolveMessage(translations, keyParts) {
```
I think this is too generic to be put inside translation.

We should preserve the safe-getter `idx`  function in `core/utils.js` to access deeply nested object and this class can instead import that utility function

The `resolveMessage` in translation just need to require this utils function

SleepWalker commented on lib/server/blog.js:
```diff
@@ -34,8 +34,12 @@ function fileToUrl(file) {
   const blogPages = {};
   const perPage = 10;
+
+  env.translation.setLanguage('en');
```
But there are translations in `<Site>` component. I think it is safer to explicitly set language to default `en`, because it may happen, that current translation language will not be 'en' due to some doc page rendered previously using other language.

SleepWalker commented on lib/server/generate.js:
```diff
@@ -386,6 +389,8 @@ async function execute() {
       const pageID = path.basename(normalizedFile, '.html');
       const parts = normalizedFile.split('pages');
       const targetFile = join(buildDir, parts[1]);
+      env.translation.setLanguage('en');
```
You are right. I'll fix it.

Can I safely remove setLanguage from `server/translate.js`? Or should I add deprecation warning there?

endiliey replied:
https://docusaurus.io/docs/en/translation#enabling-translations-on-pages

It seems safe to remove

SleepWalker commented:
> Another thing is that this will definitely clash/conflicts with #892. Whichever got merged first 

I have also finished with #868 and #827, which are improvements of sidebar too. I haven't submitted PR, because I'm coding on base of current PR.

I can wait till #892 merged and then resolve all the conflicts or if my changes are too big, I can drop them and try to resolve 868 and 827 without any refactoring.

endiliey commented:
Hey @SleepWalker,

Sorry for another late response. First of all, you might want to see https://docusaurus.io/blog/2018/09/11/Towards-Docusaurus-2

> due to work on Docusaurus 2, we will be less likely to accept new features/major changes on Docusaurus 1.

I personally prefer not to merge this since this part of docs metadata codes has been re-written & changed on v2 into pure functions without CWD/ global require anymore (no more side effects, yay !!) with > 90% test coverage. 

However, will you be interested in helping for v2 instead? I'll notify you when we've moved part of the work on this repository's v2 branch. I'm sure you'll be a great help 😃

SleepWalker commented:
Hi, @endiliey 

yep. I've read that post and it would be interesting to help with the v2 :)

I'm closing this PR

endiliey commented:
Hi @SleepWalker. I think this can now be done in v2. You might want to checkout v2 folder (although its still wip)

Feel free to ping me on GitHub or discord
